runtime/interpreter/interpreter.go

```diff
@@ -2488,6 +2489,176 @@ var fromStringFunctionValues = func() map[string]fromStringFunctionValue {
 	return values
 }()
 
+type fromBigEndianBytesFunctionValue struct {
+	receiverType sema.Type
+	hostFunction *HostFunctionValue
+}
+
+func padWithZeroes(b []byte, expectedLen int) []byte {
+	l := len(b)
+	if l > expectedLen {
+		panic(errors.NewUnreachableError())
+	} else if l == expectedLen {
+		return b
+	}
+
+	res := make([]byte, expectedLen)
+	copy(res[expectedLen-l:], b)
+	return res
+}
+
+// a function that attempts to create a Number from a big-endian bytes.
+type bigEndianBytesConverter func(*Interpreter, []byte) OptionalValue
+
+func newFromBigEndianBytesFunction(
+	ty sema.Type,
+	byteLength int,
+	converter bigEndianBytesConverter) fromBigEndianBytesFunctionValue {
+	functionType := sema.FromBigEndianBytesFunctionType(ty)
+
+	hostFunctionImpl := NewUnmeteredHostFunctionValue(
+		functionType,
+		func(invocation Invocation) Value {
+			argument, ok := invocation.Arguments[0].(*ArrayValue)
+			if !ok {
+				panic(errors.NewUnreachableError())
+			}
+
+			inter := invocation.Interpreter
+			bytes, err := ByteArrayValueToByteSlice(inter, argument, invocation.LocationRange)
+			if err != nil {
+				return Nil
+			}
+
+			// overflow
+			if byteLength != 0 && len(bytes) > byteLength {
+				return Nil
+			}
+
+			return converter(inter, bytes)
+		},
+	)
+	return fromBigEndianBytesFunctionValue{
+		receiverType: ty,
+		hostFunction: hostFunctionImpl,
+	}
+}
+
+var fromBigEndianBytesFunctionValues = func() map[string]fromBigEndianBytesFunctionValue {
+	declarations := []fromBigEndianBytesFunctionValue{
+		// signed int values
+		newFromBigEndianBytesFunction(sema.Int8Type, 1, func(i *Interpreter, b []byte) OptionalValue {
+			bytes := padWithZeroes(b, 1)
+			return NewSomeValueNonCopying(i, NewInt8Value(i, func() int8 { return int8(bytes[0]) }))
```

Comment:
Perform the allocations inside of the metered constructor, so that memory metering occurs before allocation (here and for all other functions):

```suggestion
			return NewSomeValueNonCopying(i, NewInt8Value(i, func() int8 {
				bytes := padWithZeroes(b, 1)
				return int8(bytes[0])
			}))
```

Reply:
Done except for unbounded `Int` and `UInt` types where the constructor requires the memory usage beforehand.